File: 88_Merge_Sorted_Array.py
# Leetcode problem link : https://leetcode.com/problems/merge-sorted-array/
# Time Complexity : O(n)
#  Space Complexity : O(1)
#  Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
'''
        Basic approach : O(n+m(log(n+m))) => As nums1 contains space to accommodate both so we merge both the arrays and sort. So number of elements are m+n now hence time complexity is O(n+m(log(n+m)))

        Optimized approach: O( n ) => Keep two pointers, one which is traversing the nums1 array from the end and the other traversing nums2 from the end.
        2. Current pointer will keep track of location in nums1 to be updated.
        3. Compare the elements in nums1 and nums2. Whichever is bigger is updated in nums1 at current pointer position.
        4. Decrement the current pointer and pointer of the array whose element just got udpated
'''
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # edge case when nums1 or nums2 is empty
        if not nums1 or not nums2:
            return
        # i will be pointing to last element of nums1
        i = m-1
        # j will point to last element of nums2
        j = n-1
        # current pointer will update at the end of nums1 array(as it has capacity of m+n)
        curr = m+n-1
        
        #traverse till one of them reaches the starting index
        while i>= 0 and j >=0:
            # compare and copy to current pointer position whichever is bigger, decrement relevant counters
            if nums1[i] > nums2[j]:
                nums1[curr] = nums1[i]
                i -= 1
            else:
                nums1[curr] = nums2[j]
                j -= 1
            curr -= 1
        # copy remaining elements from nums2 if left any (nums1 will anyway have its own remaining elements)
        while j >= 0:
            nums1[curr] = nums2[j]
            curr -= 1
            j -= 1

File: _88_MergeSortedArray.java
class Solution {

  // space complexity - o(1)
    //time - o(n+m)  where n = length of nums1 and m = length of nums2
    //Ran on leeetcode successfully : Yes
    // Problem faced  : No
    //Approach : Merge from the end of both the ararys, greater of two will go in the last index of nums1

  public void merge(int[] nums1, int m, int[] nums2, int n) {

       int pivot = n+m -1;
       int i = m-1;
       int j = n-1;

       while(pivot>=0){

           if(i < 0 || ( j>=0 && nums2[j]>= nums1[i])){
               nums1[pivot] = nums2[j];
               j--;
           }
           else{
               nums1[pivot] = nums1[i];
               i--;
           }
           pivot--;
       }
   }

  // space complexity - o(1)
    //time - o(n+m)
    //Ran on leeetcode successfully : Yes
    // Problem faced  : No
    //Approach : move all zeros in nums1 in front and then start merge two arrays
    public void merge(int[] nums1, int m, int[] nums2, int n) {

        int i = m-1;
        int j = n+m-1;

        while(i >= 0 ){
            nums1[j] = nums1[i];
            i--;
            j--;
        }

        i = n;
        j = 0;
        int pivot = 0;

        while(i<m+n || j<n){

            if(j>=n || ( i<m+n && nums1[i] <= nums2[j]) ){
                nums1[pivot] = nums1[i];
                i++;
            }
            else if(i>= m+n || (j < n && nums1[i] > nums2[j])){
                nums1[pivot] = nums2[j];
                j++;
            }
            pivot++;
        }

    }
}
